state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -941,23 +942,37 @@ pub mod pallet {
 							(None, None)
 						};
 
-						Some(SwapLegInfo {
-							swap_id: swap.swap_id(),
-							base_asset,
-							// All swaps to `base_asset` have to go through the stable asset:
-							quote_asset: STABLE_ASSET,
-							side: Side::Buy,
-							// Safe to unwrap as we have swapped everything into the stable asset at
-							// this point
-							amount: swap.stable_amount.unwrap(),
-							source_asset,
-							source_amount,
-						})
+						if swap.stable_amount.is_some() || pool_sell_price.is_some() {
+							// If the swap into stable asset failed, fallback to estimating the
+							// amount via pool price.
+							let amount = swap.stable_amount.unwrap_or_else(|| {
+								output_amount_ceil(
+									cf_amm::common::Amount::from(swap.input_amount()),
+									sqrt_price_to_price(
+										pool_sell_price.expect("Safe because of check above"),
```

Comment:
Not a big deal but we can probably remove a level of nesting and the `expect` by using `?` here.

```rust
let amount = swap.stable_amount.or_else(|| {
	Some(
		//...
			pool_sell_price?
		//
	)
})?;

Reply:
I tried 3 versions of this logic trying to make it better, and never thought of that. 😅